### PR TITLE
Refactor decision tree classification as classed R bindings

### PR DESCRIPTION
**Pull request by eddelbuettel**
This PR follow earlier PRs and refactors the decision tree classifier into a classed binding for R along with new tests.

### CODE REVIEW

File: src/mlpack/methods/decision_tree/decision_tree_classify_main.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+// Short description.
+BINDING_SHORT_DESC("Class predictions from train decision tree model.");
+
+// Long description.
+BINDING_LONG_DESC("")
+
+// Example.
+BINDING_EXAMPLE(
+    CALL_METHOD("model", "classify", "test", "X_test"));
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_classify_main.cpp:37]: I think this needs to be `METHOD_CALL()` and that will fix the build (or at least bring it closer to fixed :)).

File: src/mlpack/methods/decision_tree/decision_tree_train_main.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+BINDING_SEE_ALSO("Decision trees on Wikipedia",
+    "https://en.wikipedia.org/wiki/Decision_tree_learning");
+BINDING_SEE_ALSO("Induction of Decision Trees (pdf)",
+    "https://www.hunch.net/~coms-4771/quinlan.pdf");
+BINDING_SEE_ALSO("DecisionTree C++ class documentation",
+    "@doc/user/methods/decision_tree.md");
+
+// Datasets.
+PARAM_MATRIX_IN_REQ("training", "Training dataset (may be categorical).",
+    "t");
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]: We really need to keep this as `MATRIX_AND_INFO`, otherwise all the claims about support for categorical data are incorrect, and the new bindings will lack a pretty critical feature (the ability to support categorical datasets).  So I know it will be potentially a little bit painful, but can we change it back and sort out the failures?  I'm happy to help with any debugging pain.  I believe none of the grouped bindings have a matrix+info parameter yet, so there might be some pieces of support that are yet to be implemented.

File: src/mlpack/methods/decision_tree/decision_tree_train_main.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+// Models.
+PARAM_MODEL_OUT(DecisionTreeModel, "output_model", "Output for trained decision"
+    " tree.", "M");
+
+void BINDING_FUNCTION(util::Params& params, util::Timers& /* timers */)
+{
+  // Check parameters.
+  ReportIgnoredParam(params, {{ "training", false }},
+      "print_training_accuracy");
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]: Much like in the other PR, this check can no longer fail, so we can remove it.

**Review by rcurtin — COMMENTED** [review]
The general structure of course looks just fine; the extra matrix+info parameter might throw us for a little bit of a loop but I'm sure we can work through it without *too* much pain :)

File: src/mlpack/methods/decision_tree/decision_tree_train_main.cpp (hunk 2 above)
Comment (eddelbuettel) [reply src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]: We can chat Monday.  Givem that `MATRIX_AND_INFO` is a two data structures, I do no see how we can map this to R's interface.  The best idea I had so far would be to add a flag, and if set go use the (to extraced code from the reader) to fill DatasetInfo from parsing the matrix passed in.

File: src/mlpack/methods/decision_tree/decision_tree_classify_main.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+// Datasets.
+PARAM_MATRIX_AND_INFO_IN_REQ("test", "Testing dataset (may contain "
+    "categorical variables).", "T");
+PARAM_UROW_IN("test_labels", "Test point labels, if accuracy calculation "
+    "is desired.", "L");
+
+// Output parameters.
+//PARAM_MATRIX_OUT("probabilities", "Class probabilities for each test point "
+//    "if probabilities has been selected.", "P");
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_classify_main.cpp]: I think it would be fine to just remove the commented-out `probabilities` line here.

File: src/mlpack/tests/ann/layer_test.cpp [hunk 5]
```diff
@@ -15,16 +15,16 @@
...
+//  #define MLPACK_ENABLE_ANN_SERIALIZATION
+//#endif
 
 #include "layer/adaptive_max_pooling.cpp"
 #include "layer/adaptive_mean_pooling.cpp"
 #include "layer/add.cpp"
 #include "layer/add_merge.cpp"
 #include "layer/alpha_dropout.cpp"
-#include "layer/batch_norm.cpp"
+//#include "layer/batch_norm.cpp"
```
Comment (rcurtin) [inline src/mlpack/tests/ann/layer_test.cpp]: Thank you for debugging this.  I can see from the CI history that this strategy does work, but, I don't want to totally remove tests for the batch normalization layer.  I believe that the following will work:

 * Re-enable `batch_norm.cpp` here
 * Comment out the tests `BatchNormSerializationTest` in `layer/batch_norm.cpp`

File: src/mlpack/tests/ann/layer_test.cpp [hunk 6]
```diff
@@ -15,16 +15,16 @@
  * 3-clause BSD license along with mlpack.  If not, see
  * http://www.opensource.org/licenses/BSD-3-Clause for more information.
  */
-#ifndef MLPACK_ENABLE_ANN_SERIALIZATION
-  #define MLPACK_ENABLE_ANN_SERIALIZATION
-#endif
+//#ifndef MLPACK_ENABLE_ANN_SERIALIZATION
+//  #define MLPACK_ENABLE_ANN_SERIALIZATION
+//#endif
```
Comment (rcurtin) [inline src/mlpack/tests/ann/layer_test.cpp:18]:
```suggestion

// NOTE: serialization is not enabled for individual layer tests, in order to
// reduce compilation time and memory usage.  Any tests for serialization of
// layers or networks should be done separately.
```

What do you think?  It's probably good to add a comment like this here.

File: src/mlpack/tests/main_tests/decision_tree_train_test.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+TEST_CASE_METHOD(DecisionTreeTrainTestFixture,
+                 "DecisionTreeTrainOutputDimensionTest",
+                 "[DecisionTreeTrainMainTest][BindingTests]")
+{
+  constexpr int N = 10;
+  constexpr int D = 2;
+  arma::mat trainX = arma::trans(arma::randu<arma::mat>(N, D));
+  arma::Row<size_t> trainY = arma::randu<arma::Row<size_t>>(N);
+  TextOptions opts;
+  opts.Categorical() = true;
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/decision_tree_train_test.cpp]: You could alternately just make a `DatasetInfo` object here:

```
DatasetInfo info(D);
```

and then pass that to `std::make_tuple()`.

(More documentation on `DatasetInfo` [here](https://www.mlpack.org/doc/user/load_save.html#datasetinfo).)

File: src/mlpack/tests/main_tests/decision_tree_train_test.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  SetInputParam("minimum_gain_split", (double) -0.1);   // Invalid
+  REQUIRE_THROWS_AS(RUN_BINDING(), std::runtime_error);
+
+  SetInputParam("minimum_gain_split", (double) 1.1);    // Invalid
+  REQUIRE_THROWS_AS(RUN_BINDING(), std::runtime_error);
+
+  SetInputParam("minimum_gain_split", (double) 1e-7);   // Now valid.
+  SetInputParam("maximum_depth", (int) -1);             // Invalid.
+  REQUIRE_THROWS_AS(RUN_BINDING(), std::runtime_error);
+}
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/decision_tree_train_test.cpp:104]: Do you want to additionally test this on a matrix that has a categorical feature?  Here is how you would do that... or one example that you could adapt:

```
arma::mat trainX = arma::randu<arma::mat>(D, N);
DatasetInfo info(d);

// Make last row categorical with 5 categories.
trainX.row(D - 1) = arma::randi<arma::rowvec>(N, arma::distr_param(0, 4));
info.Type(D - 1) = Datatype::categorical;
```

I *think* that's all that's necessary.  If that gives a compilation failure or something else let me know, there may be one thing else to do to set up the `DatasetInfo`.

File: src/mlpack/tests/CMakeLists.txt [hunk 9]
```diff
@@ -208,9 +211,6 @@ add_executable(mlpack_test
   main_tests/preprocess_split_test.cpp
   main_tests/radical_test.cpp
   main_tests/random_forest_test.cpp
-  main_tests/random_forest_train_test.cpp
-  main_tests/random_forest_classify_test.cpp
-  main_tests/random_forest_probabilities_test.cpp
```
Comment (rcurtin) [inline src/mlpack/tests/CMakeLists.txt:216]: Accidental commit?

File: src/mlpack/methods/decision_tree/decision_tree_classify_main.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+      if (predictions[i] == testLabels[i])
+        ++correct;
+
+    // Print number of correct points.
+    Log::Info << double(correct) / double(testPoints.n_cols) * 100 << "% "
+      << "correct on test set (" << correct << " / " << testPoints.n_cols
+      << ")." << endl;
+  }
+
+  params.Get<arma::Row<size_t>>("predictions") = predictions;
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_classify_main.cpp]:
```suggestion
  params.Get<arma::Row<size_t>>("predictions") = std::move(predictions);
```
Could save a copy.  Of trivial importance.

File: src/mlpack/methods/decision_tree/decision_tree_probabilities_main.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+      if (predictions[i] == testLabels[i])
+        ++correct;
+
+    // Print number of correct points.
+    Log::Info << double(correct) / double(testPoints.n_cols) * 100 << "% "
+      << "correct on test set (" << correct << " / " << testPoints.n_cols
+      << ")." << endl;
+  }
+
+  params.Get<arma::mat>("probabilities") = probabilities;
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_probabilities_main.cpp]:
```suggestion
  params.Get<arma::mat>("probabilities") = std::move(probabilities);
```

**Review by rcurtin — COMMENTED** [review]
Thanks for digging in on this one and debugging the unrelated CI failure.  I know you didn't ask for that problem and it has nothing to do with the changes you are actually working on here...

Functionality-wise everything looks good; I have a handful of small suggestions, hopefully they are helpful.  I do want to make sure we re-enable the batch norm layer tests so much as we reasonably can before merge.

File: src/mlpack/methods/decision_tree/decision_tree_train_main.cpp [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    }
+    else
+    {
+      model->tree.Train(std::move(trainingSet), std::move(labels), numClasses,
+          minLeafSize, minimumGainSplit, maxDepth);
+    }
+  }
+
+  // Do we need to print training error?
+  if (print_training_accuracy)
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]:
```suggestion
  if (params.Has("print_training_accuracy"))
```

There's a little inconsistency throughout... once, `print_training_accuracy` gets used (like here), in other places, `params.Has("print_training_accuracy")` is used.  Personally?  I'd just use `params.Has()` throughout and remove the variable (not part of my suggestion here).  But if you want to use the variable instead, that works too.  This is a really minor comment... the code works just fine either way.

**Review by rcurtin — APPROVED** [review]
Everything looks good to me here!  Great to have the dataset+info parameters working correctly.  Thanks again for digging into the unrelated CI issue and working that out.

File: src/mlpack/methods/decision_tree/decision_tree_train_main.cpp (hunk 12 above)
Comment (eddelbuettel) [reply src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]: We have

https://github.com/eddelbuettel/mlpack/blob/1148b8369d2764a40dc2bc4f0bfa8fa6ceaf856c/src/mlpack/methods/decision_tree/decision_tree_train_main.cpp#L114

which boo-hoo does not expand here but is

```c++
bool print_training_accuracy = params.Has("print_training_accuracy");
```

and the pattern of turning a parameter flag into a boolean appears in multiple files.  ~I could remove the single use boolean, I suspect the compiler might do that for us anyway.~  Corrected, it is not single-use as lines ~149~ 162 and 175 rely on it.

Comment (rcurtin) [reply src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]: May as well change the `params.Has("print_training_accuracy")` uses then into uses of the variable?  This is a really minor point either way.  Your call on what to do.

Comment (eddelbuettel) [reply src/mlpack/methods/decision_tree/decision_tree_train_main.cpp]: Indeed, I overlooked the 3rd use on line 149.  I now removed the `bool`; all three uses acccess `params`.
